fcrepo4/fcrepo4
# Allow single-document ACLs using hash URIs for authorizations.

Resolves: https://jira.duraspace.org/browse/FCREPO-2742

**Allow single-document ACLs using hash URIs for authorizations.**
* * *

**JIRA Ticket**: https://jira.duraspace.org/browse/FCREPO-2742
 

* Other Relevant Links (Mailing list discussion, related pull requests, etc.)

# What does this Pull Request do?
This PR changes the algorithm for resolving the set of roles that applies to a given resource, bringing it into line with the SOLID/WebAC specification.   All unit and integration tests that were 

# What's new?
Now ACLs defined at /<resource path>/fcr:acl  are applied.  If no /fcr:acl is defined, fedora will search up the ancestor hierarchy until it finds an ACL. 

# How should this be tested?

There are quite a number of integration tests in place which you can look at for inspiration.  But for starters, you can now create resource, create an ACL for that resource and verify that the resource is protected.  Here's an example: 

1. create the file acl.ttl with the following contents
```
@prefix acl: <http://www.w3.org/ns/auth/acl#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .

<#restricted> a acl:Authorization ;
   acl:agent "testuser" ;
   acl:mode acl:Read ;
   acl:accessTo </rest/test/restricted> .
```
2. create two resources: 
```
curl -u fedoraAdmin:fedoraAdmin -X PUT http://localhost:8080/rest/test
curl -u fedoraAdmin:fedoraAdmin -X PUT http://localhost:8080/rest/test/restricted
```
3. create the acl
```
curl -u fedoraAdmin:fedoraAdmin -v -X PUT  http://localhost:8080/rest/test/fcr:acl --data-binary "@acl.ttl" -H "Content-Type: text/turtle"
```
4. verify that open is open and restricted is restricted
```
# any authenticated user should be able to access
curl -u testuser2:testpass -X GET http://localhost:8080/rest/test/restricted 
#should return 403
curl -u testuser:testuser -X GET http://localhost:8080/rest/test/restricted
#should return 200
```

# Additional Notes:
# Interested parties
@awoods, @peichman-umd , @fcrepo4/committers
